In [98]:
import pandas as pd
import numpy as np
from datetime import datetime,timedelta
import seaborn as sns
from sklearn.preprocessing import StandardScaler

In [27]:
sns.heatmap(data.corr())

## Prepro

In [331]:
#df = pd.read_csv("train_data_Akhmet_v2.csv")
df = pd.read_csv("2_percent.csv")
print(df.shape)

(150643, 89)


In [314]:
df['cancelled'].value_counts()

0.0    147883
1.0      2760
Name: cancelled, dtype: int64

In [315]:
#dfx = pd.read_csv("2_percent.csv")
#print(dfx.shape)

(150643, 89)


In [303]:
dfx['cancelled'].value_counts()

0.0    147883
1.0      2760
Name: cancelled, dtype: int64

In [332]:
df=df[['mkt_unique_carrier', 'op_unique_carrier',
       'op_carrier_fl_num', 'arr_delay', 'crs_elapsed_time', 'distance',
       'inbound_flights', 'arr_month', 'arr_day_of_week',
       'arr_hour', 'dep_month', 'dep_day_of_week', 'dep_hour', 'hub_type_or',
       'area_or', 'total_operations_annual_or', 'hub_type_des',
       'facility_use_des', 'area_des', 'total_operations_annual_des',
       'origin_state', 'dest_state', 'cancelled', 'nas_delay',
       'security_delay', 'weather_delay', 'carrier_delay',
       'late_aircraft_delay','num_runways_des',
       'avg_runway_length_des', 'avg_runway_width_des','arr_type', 'arr_severity', 'dep_type', 'dep_severity']]
print(df.shape)
print(df.isnull().sum())

(150643, 35)
mkt_unique_carrier                  0
op_unique_carrier                   0
op_carrier_fl_num                   0
arr_delay                        3155
crs_elapsed_time                    0
distance                            0
inbound_flights                     0
arr_month                           0
arr_day_of_week                     0
arr_hour                            0
dep_month                           0
dep_day_of_week                     0
dep_hour                            0
hub_type_or                        48
area_or                           182
total_operations_annual_or          0
hub_type_des                       34
facility_use_des                    0
area_des                          174
total_operations_annual_des         0
origin_state                        0
dest_state                          0
cancelled                           0
nas_delay                      122423
security_delay                 122423
weather_delay                  122423

In [333]:
df.drop(columns=['nas_delay',
       'security_delay', 'weather_delay', 'carrier_delay',
       'late_aircraft_delay'],inplace=True)

In [334]:
#df=df.drop(columns=['arr_month', 'arr_day_of_week',
       #'arr_hour', 'dep_month', 'dep_day_of_week', 'dep_hour'])

In [340]:
#df.dropna(inplace=True)
df.fillna(0,inplace=True)
print(df.shape)

(150643, 30)


In [341]:
print(df.isnull().sum())

mkt_unique_carrier             0
op_unique_carrier              0
op_carrier_fl_num              0
arr_delay                      0
crs_elapsed_time               0
distance                       0
inbound_flights                0
arr_month                      0
arr_day_of_week                0
arr_hour                       0
dep_month                      0
dep_day_of_week                0
dep_hour                       0
hub_type_or                    0
area_or                        0
total_operations_annual_or     0
hub_type_des                   0
facility_use_des               0
area_des                       0
total_operations_annual_des    0
origin_state                   0
dest_state                     0
cancelled                      0
num_runways_des                0
avg_runway_length_des          0
avg_runway_width_des           0
arr_type                       0
arr_severity                   0
dep_type                       0
dep_severity                   0
dtype: int

In [342]:
X,y= df.drop(columns=['cancelled',"arr_delay"]),df[['cancelled']]

X_dummy = X.drop(columns=["op_carrier_fl_num","crs_elapsed_time",])

In [343]:
X.columns

Index(['mkt_unique_carrier', 'op_unique_carrier', 'op_carrier_fl_num',
       'crs_elapsed_time', 'distance', 'inbound_flights', 'arr_month',
       'arr_day_of_week', 'arr_hour', 'dep_month', 'dep_day_of_week',
       'dep_hour', 'hub_type_or', 'area_or', 'total_operations_annual_or',
       'hub_type_des', 'facility_use_des', 'area_des',
       'total_operations_annual_des', 'origin_state', 'dest_state',
       'num_runways_des', 'avg_runway_length_des', 'avg_runway_width_des',
       'arr_type', 'arr_severity', 'dep_type', 'dep_severity'],
      dtype='object')

In [344]:
y['cancelled'].value_counts()
#ya= df[['nas_delay',
      # 'security_delay', 'weather_delay', 'carrier_delay',
       #'late_aircraft_delay']]
#yb= df['cancelled']
#ya=ya.fillna(0)

0.0    147883
1.0      2760
Name: cancelled, dtype: int64

In [208]:
y['type']=y.apply(lambda x: x.idxmax(),axis=1)

<ipython-input-208-d6f83c156f4d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y['type']=y.apply(lambda x: x.idxmax(),axis=1)


In [209]:
y['type'].replace({"nas_delay": 0, "security_delay": 1,"weather_delay": 2,"carrier_delay": 3,"late_aircraft_delay": 4}, inplace=True)

C:\Users\tokis\anaconda3\lib\site-packages\pandas\core\generic.py:6746: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [346]:
#y=np.asarray(y['type'])
y=np.asarray(y['cancelled'])
print(y.shape)
print(X.shape)
print(X_dummy.shape)

(150643,)
(150643, 28)
(150643, 26)


In [347]:
X_dummy = pd.get_dummies(
    data=X_dummy,columns=["mkt_unique_carrier","op_unique_carrier",
                        "hub_type_or",
                          "hub_type_des",
                        "facility_use_des",
                        "origin_state","dest_state",'arr_type', 'arr_severity', 'dep_type', 'dep_severity',
                          ])
X_dummy=StandardScaler().fit_transform(X_dummy)
X_dummy=np.asarray(X_dummy)


In [348]:
#"arr_month","arr_day_of_week","arr_hour","dep_month","dep_hour","dep_day_of_week"

In [349]:
print(np.mean(y))
print(np.median(y))

0.01832146199956188
0.0


In [350]:
print(y.shape)
print(X_dummy.shape)

(150643,)
(150643, 199)


AttributeError: 'numpy.ndarray' object has no attribute 'isnull'

from sklearn.decomposition import PCA
import plotly.express as px

X = X_dummy

pca = PCA(n_components=10)
principalComponents = pca.fit_transform(X)
print(principalComponents.shape)
print(type(principalComponents.shape))

In [216]:
#principalComponents=np.asarray(principalComponents)

## Modelling

from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import r2_score,mean_absolute_error,mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from imblearn.over_sampling import RandomOverSampler
from imblearn.over_sampling import SMOTE
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier


Multiclass Classification
The target variables are CARRIER_DELAY, WEATHER_DELAY, NAS_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY. We need to do additional transformations because these variables are not binary but continuos. For each flight that was delayed, we need to have one of these variables as 1 and others 0.
It can happen that we have two types of delays with more than 0 minutes. In this case, take the bigger one as 1 and others as 0.

In [351]:
oversample = SMOTE(sampling_strategy='minority')
#oversample = SMOTE(sampling_strategy='all')
X1, y1 = oversample.fit_resample(X_dummy, y)

In [352]:
print(X1.shape)
print(y1.shape)

(295766, 199)
(295766,)


In [353]:
unique, counts = np.unique(y1, return_counts=True)
dict(zip(unique, counts))

{0.0: 147883, 1.0: 147883}

In [357]:
X_train,X_test,y_train,y_test = train_test_split(X1,y1,train_size=0.7,random_state=0)

In [358]:
#clf = LogisticRegression(random_state=0)
#clf = LogisticRegression(random_state=0,max_iter=20000, multi_class='multinomial')

#clf = KNeighborsClassifier(n_neighbors=5)

#clf = GaussianNB()

#clf = RandomForestClassifier(max_depth=5, random_state=0)

#clf = SVC(kernel='poly',gamma='auto')

In [370]:
clf = LogisticRegression(random_state=0,max_iter=20000, multi_class='multinomial')
clf.fit(X_train, y_train)
y_pred=clf.predict(X_test)


In [371]:
print(type(y_test))
print(type(y_pred))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [377]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix
accuracy = accuracy_score(y_test,y_pred)
f1_score = f1_score(y_test,y_pred,average='weighted')
precision_score = precision_score(y_test,y_pred,average='weighted')
recall_score = recall_score(y_test,y_pred,average='micro')
cnf_matrix = confusion_matrix(y_test, y_pred)

#### Multi Logistic Regression SMOTE without Scaler
0.28330522765598654
0.25589846716567505
0.2746703349064258
0.28330522765598654
[[ 262  525  160  145  423]
 [ 255 1025   42   23  215]
 [ 225  525  211  248  342]
 [ 197  609  167  224  349]
 [ 182  573  149  171  462]]
 
#### Multi Logistic Regression SMOTE with Scaler
0.28330522765598654
0.25589846716567505
0.2746703349064258
0.28330522765598654
[[ 262  525  160  145  423]
 [ 255 1025   42   23  215]
 [ 225  525  211  248  342]
 [ 197  609  167  224  349]
 [ 182  573  149  171  462]]
 PCA MAKES THINGS WORST 10=33%Accuracy 100=45%


In [378]:
print(accuracy)
print(f1_score)
print(precision_score)
print(recall_score)
print(cnf_matrix)

0.9906006987490139
0.9906000634710523
0.990774509024691
0.9906006987490139
[[44267     0]
 [  834 43629]]


In [376]:
import xgboost as xgb
data_dmatrix = xgb.DMatrix(data=X_train,label=y_train)
xg_reg = xgb.XGBRegressor(objective ='binary:hinge', colsample_bytree = 0.3, learning_rate = 0.1,
                max_depth = 2, alpha = 0.1, n_estimators = 1000)
xg_reg.fit(X_train,y_train)
y_pred = xg_reg.predict(X_test)

In [253]:
from sklearn.ensemble import RandomForestRegressor

In [203]:
from sklearn import metrics
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import GradientBoostingRegressor
param_grid={
    'alpha' :[0.1,0.3,0.5,0.7,1.0],
    'max_depth': [1,3,5,7,10],
    'n_estimators': [10,100,250,500,1000]
    }
params = {
    # Parameters that we are going to tune.
    'max_depth':6,
    'min_child_weight': 1,
    'eta':.3,
    'subsample': 1,
    'colsample_bytree': 1,
    # Other parameters
    'objective':'reg:squarederror',
}

## ModellingII

0.43279074906091186
[[1999  277  285   13   98]
 [  73 2646   55    0    0]
 [ 979  646 1038    0   18]
 [1493  717  397   19  104]
 [1928  452  154   12  174]]


XGBRegressor(alpha=0.1, base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.3, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.1, max_delta_step=0, max_depth=2,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=1000, n_jobs=16, num_parallel_tree=1,
             objective='binary:hinge', random_state=0, reg_alpha=0.100000001,
             reg_lambda=1, scale_pos_weight=None, subsample=1,
             tree_method='exact', validate_parameters=1, verbosity=None)